compiler/src/dotty/tools/dotc/transform/SpecializeFunction1.scala

```diff
@@ -0,0 +1,164 @@
+
+    Function1 = ctx.requiredClass("scala.Function1")
+
+    blacklisted = Set(
+      "scala.compat.java8.JFunction1",
+      "scala.runtime.AbstractFunction1",
+      "scala.PartialFunction",
+      "scala.runtime.AbstractPartialFunction"
+    ).map(ctx.requiredClass(_))
+
+    this
+  }
+
+  private object ShouldTransformDenot {
+    def unapply(cref: ClassDenotation)(implicit ctx: Context): Option[(ClassDenotation, Type, Type, Type)] = {
+      def collectFunc1(xs: List[Type])(implicit ctx: Context): Option[(Type, Type, Type)] =
+        xs.collect {
+          case func1 @ RefinedType(RefinedType(parent, _, t1), _, r)
+            if func1.isRef(Function1) => (t1, r, func1)
+        }.headOption
+
+      collectFunc1(cref.info.parentsWithArgs).flatMap { case (t1, r, func1) =>
+        if (
+          !argTypes.contains(t1.typeSymbol) ||
+          !returnTypes.contains(r.typeSymbol) ||
+          blacklisted.exists(sym => cref.symbol.derivesFrom(sym))
+        ) None
+        else Some((cref, t1, r, func1))
+      }
+    }
+  }
+
+  private object ShouldTransformTree {
+    def unapply(tree: Template)(implicit ctx: Context): Option[(Tree, Type, Type)] =
+      tree.parents
+      .map { t => (t.tpe, t) }
+      .collect {
+        case (tp @ RefinedType(RefinedType(parent, _, t1), _, r), func1)
+        if tp.isRef(Function1) &&
+           argTypes.contains(t1.typeSymbol) &&
+           returnTypes.contains(r.typeSymbol) => (func1, t1, r)
+      }
+      .headOption
+  }
+
+  private def specializedName(name: Name, t1: Type, r: Type)(implicit ctx: Context): Name =
+    name.specializedFor(List(t1, r), List(t1, r).map(_.typeSymbol.name))
+
+  def transformDenot(cref: ClassDenotation, t1: Type, r: Type, func1: Type)(implicit ctx: Context): SingleDenotation = {
+    val specializedFunction: TypeRef =
+      ctx.requiredClassRef(functionPkg ++ specializedName(functionName, t1, r))
+
+    def replaceFunction1(in: List[TypeRef]): List[TypeRef] =
+      in.foldRight(List.empty[TypeRef]) { (tp, acc) =>
+        val newTp =
+          if (tp.isRef(Function1)) specializedFunction
+          else tp
+        newTp :: acc
+      }
+
+    def specializeApply(scope: Scope): Scope = {
+      def specializedApply: Symbol = {
+        val specializedMethodName = specializedName(nme.apply, t1, r)
+        ctx.newSymbol(
+          cref.symbol,
+          specializedMethodName,
+          Flags.Override | Flags.Method,
+          specializedFunction.info.decls.lookup(specializedMethodName).info
+        )
+      }
+
+      val alteredScope = scope.cloneScope
+      alteredScope.enter(specializedApply)
+      alteredScope
+    }
+
+    val ClassInfo(prefix, cls, parents, decls, info) = cref.classInfo
+    val newInfo = ClassInfo(prefix, cls, replaceFunction1(in = parents), specializeApply(decls), info)
+    cref.copySymDenotation(info = newInfo)
+  }
+
+  private def transformTree(tmpl: Template, func1: Tree, t1: Type, r: Type)(implicit ctx: Context) = {
```

Comment:
I wanted to keep things apart so that it would be easier to get a quick overview of the workings of `SpecializeFunction1`, but perhaps it was taking it a bit far...